cmd: Add `cilium map history` command

vadorovsky opened the pull request:
This change introduces a command with the following syntax:

```
cilium map get <name> history
```

It shows last 50 operations on the given BPF map.

Fixes: #8179

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8253)
<!-- Reviewable:end -->

vadorovsky committed:
bpf: Add history of BPF map operations

This change introduces a storage for BPF map operations history which
stores last 50 operations with timestamp.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
api: Add BPF map history endpoint and models

This change prepares API endpoint and models for BPF map operations
history.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
api: Add auto-generated files

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
daemon: Add API handler for BPF map history

This change adds the API handler for BPF map operations history.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
cmd: Add `cilium map history` command

This change introduces a command with the following syntax:

```
cilium map history <name>
```

It shows last 50 operations on the given BPF map.

Fixes: #8179

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

aanm commented on pkg/bpf/map_linux.go:
```diff
@@ -88,6 +93,56 @@ type cacheEntry struct {
+	entries []*historyEntry
+}
+
+func (hm *HistoryManager) addEntry(key MapKey, value MapValue,
```
Please add a benchmark to test the memory and CPU consumption of this function.

- `time.Now()` will be extremely costly to be done on a per operation basis.

vadorovsky replied:
I added benchmark tests for `addEntry` and `addDeleteAllEntry`. The results are:

```
 mrostecki   bpf-map-history  …  github.com  cilium  cilium  make bench TESTPKGS="pkg/bpf" CONTAINER_ENGINE="sudo podman"
can't load package: package github.com/cilium/cilium/test/bpf: C source files not allowed when not using cgo or SWIG: bpf-event-test.c elf-demo.c unit-test.c
Starting key-value store containers...
sudo podman rm -f "cilium-etcd-test-container" 2> /dev/null
[sudo] password for root: 
make: [Makefile:72: start-kvstores] Error 1 (ignored)
sudo podman run -d \
	--name "cilium-etcd-test-container" \
	-p 4002:4001 \
	quay.io/coreos/etcd:v3.2.17 \
	etcd -name etcd0 \
	-advertise-client-urls http://0.0.0.0:4001 \
	-listen-client-urls http://0.0.0.0:4001 \
	-listen-peer-urls http://0.0.0.0:2380 \
	-initial-cluster-token etcd-cluster-1 \
	-initial-cluster-state new
4e81ec74535ee586535e246d2c4d2488968813f21c391b2ced0ce333b59278aa
sudo podman rm -f "cilium-consul-test-container" 2> /dev/null
make: [Makefile:74: start-kvstores] Error 1 (ignored)
rm -rf /tmp/cilium-consul-certs
mkdir /tmp/cilium-consul-certs
cp /home/mrostecki/go/src/github.com/cilium/cilium/test/consul/* /tmp/cilium-consul-certs
chmod -R a+rX /tmp/cilium-consul-certs
sudo podman run -d \
	--name "cilium-consul-test-container" \
	-p 8501:8443 \
	-e 'CONSUL_LOCAL_CONFIG={"skip_leave_on_interrupt": true, "disable_update_check": true}' \
	-v /tmp/cilium-consul-certs:/cilium-consul/ \
	consul:1.1.0 \
	agent -client=0.0.0.0 -server -bootstrap-expect 1 -config-file=/cilium-consul/consul-config.json
b6b8bd206331292e610e7230fe59bb829042e78e4ee83197b20973c8e09bac6c
go test -ldflags "-X github.com/cilium/cilium/pkg/kvstore.consulDummyConfigFile=/tmp/cilium-consul-certs/cilium-consul.yaml -X github.com/cilium/cilium/pkg/testutils.CiliumRootDir=/home/mrostecki/go/src/github.com/cilium/cilium -X github.com/cilium/cilium/pkg/datapath.DatapathSHA=1234567890abcdef7890" -test.v -timeout 360s -bench="." -run=^-benchtime=10s github.com/cilium/cilium/pkg/bpf || exit 1;
PASS
ok  	github.com/cilium/cilium/pkg/bpf	0.036s
make stop-kvstores
make[1]: Entering directory '/home/mrostecki/go/src/github.com/cilium/cilium'
can't load package: package github.com/cilium/cilium/test/bpf: C source files not allowed when not using cgo or SWIG: bpf-event-test.c elf-demo.c unit-test.c
sudo podman rm -f "cilium-etcd-test-container"
4e81ec74535ee586535e246d2c4d2488968813f21c391b2ced0ce333b59278aa
sudo podman rm -f "cilium-consul-test-container"
b6b8bd206331292e610e7230fe59bb829042e78e4ee83197b20973c8e09bac6c
rm -rf /tmp/cilium-consul-certs
make[1]: Leaving directory '/home/mrostecki/go/src/github.com/cilium/cilium'
```

What do you suggest as an alternative? Getting rid of timestamp? Having a timestamp updated each few/dozens of seconds in some goroutime and then using it in `addEntry`/`addDeleteAllEntry`?

brb replied:
I guess it makes sense to add benchmarking to `*Map.Update()` and `*Map.Delete()`, and to see what's the difference between with and without the history manager.

Anyway, for those who care a lot about performance, maybe it makes sense to add an option to disable the history manager?

aanm commented on pkg/bpf/map_linux.go:
```diff
@@ -88,6 +93,56 @@ type cacheEntry struct {
+		return
+	}
+	hm.i++
+	hm.entries = append(hm.entries, entry)
```
Is there a protection against concurrent access of this slide?

vadorovsky replied:
Sorry for that! I added the RWMutex.

aanm commented on pkg/bpf/map_linux.go:
```diff
@@ -908,6 +969,8 @@ func (m *Map) DeleteAll() error {
 			log.WithError(err2).Warningf("Unable to correlate iteration key %v with cache entry. Inconsistent cache.", nextKey)
 		}
 
+		defer m.History.addDeleteEntry(mk, err)
```
It would be better to perform some kind of aggregation. If the map has 10000 entries, we will be calling `addDeleteEntry` 10000 times although we only care the last 50. The aggregation could be done on a per `DeleteAll()` function call where we would store the number of entries deleted and store it in the `m.History`.

vadorovsky replied:
I have (probably) even a better idea - when we delete all entries in the map, maybe let's just leave it as one information with key value `ALL`? Which would mean that we performed deletion of all keys.

I pushed that change. Let me know if that approach is OK for you.

vadorovsky committed:
bpf: Rename BPF map privileged tests module

Next changes are going to introduce non-privileged BPF map tests, both
kind of test suites need to be distinguished and kept in separate files.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
Documentation: Update cmdref

Auto-generated update of cmdref.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

joestringer commented on api/v1/openapi.yaml:
```diff
@@ -680,6 +680,20 @@ paths:
+      - daemon
+      parameters:
+      - "$ref": "#/parameters/map-name"
+      responses:
```
Depending on the cost of this feature, it might be useful to have a daemon option to disable it and have another error code here that represents that the feature is disabled.

aanm replied:
Maybe we should enable it based if debug is set or not.

joestringer replied:
I'm afraid that if you tie it to debug, then it will never be used. IMO it should be on by default but able to be turned off.

joestringer commented on Documentation/cmdref/cilium_map.md:
```diff
@@ -26,5 +26,6 @@ Access BPF maps
 
 * [cilium](../cilium)	 - CLI
 * [cilium map get](../cilium_map_get)	 - Display BPF map information
+* [cilium map history](../cilium_map_history)	 - Display BPF map operations history
```
Please also add this tool to `bugtool` so that we gather this in sysdumps.

joestringer reviewed (changes requested):
Couple of minor nits. Could you also show the regular, json output examples in the PR description?

brb commented on pkg/bpf/map_linux.go:
```diff
@@ -37,10 +37,16 @@ import (
 )
 
+const (
+	maxHistoryEntries = 50
```
Perhaps this can be configurable? Could be done in a separate PR to keep this PR size sane.

vadorovsky replied:
Done.

brb commented on pkg/bpf/map_linux.go:
```diff
@@ -88,6 +94,94 @@ type cacheEntry struct {
+type historyEntry struct {
+	Timestamp time.Time
+	Key       MapKey
+	AllKeys   bool
```
I would add a small comment along the lines of `AllKeys denotes whether Action was performed on all map entries`.

brb commented on pkg/bpf/map_linux.go:
```diff
@@ -88,6 +94,94 @@ type cacheEntry struct {
+
+	hm.mutex.Lock()
+	defer hm.mutex.Unlock()
+	hm.entries = append(hm.entries, entry)
```
We probably want to do the same bookkeeping here to avoid exceeding max entries?

vadorovsky replied:
Thanks for pointing that out. :man_facepalming: 

I moved the whole bookkeeping logic to ´addEntryMeta` and all the other methods are calling it.

brb commented on pkg/bpf/map_linux.go:
```diff
@@ -88,6 +94,94 @@ type cacheEntry struct {
+	hm.entries = append(hm.entries, entry)
+}
+
+func (hm *HistoryManager) addDeleteAllEntry(err error) {
```
I think it makes sense to add `deleteEntry` method which would be called by `*Map.Delete()`?

vadorovsky replied:
Done.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -1977,6 +1991,39 @@ definitions:
+        description: Action which was performed
+        type: string
+        enum:
+        - ok
```
What does it mean the "Ok" operation?

vadorovsky replied:
https://github.com/cilium/cilium/blob/master/pkg/bpf/map.go#L138-L140

aanm commented on pkg/bpf/map_linux.go:
```diff
@@ -88,6 +94,102 @@ type cacheEntry struct {
+// HistoryManager is an utility which stores history of operations on the BPF
+// map.
+type HistoryManager struct {
+	i       int
```
What is `i`?

vadorovsky replied:
The number of already stored elements.

aanm commented on pkg/bpf/map_linux.go:
```diff
@@ -88,6 +94,102 @@ type cacheEntry struct {
+	hm.mutex.Lock()
+	defer hm.mutex.Unlock()
+
+	// It max amount of entries is exceeded, get the subslice without the
```
You have already allocated the slice so you can simply do.
```golang
hm.entries[i] = entry
i++
if i > maxEntries {
i = 0
}
```

vadorovsky replied:
True, good idea!

aanm commented on pkg/bpf/map_linux.go:
```diff
@@ -881,15 +998,15 @@ func (m *Map) DeleteAll() error {
 	}
 
-	if err := m.Open(); err != nil {
+	if err = m.Open(); err != nil {
```
non-blocking but the `err :=` -> `err = ` can be done a different commit / PR

vadorovsky replied:
The only reason why I did `err :=` -> `err =` is that I want to reuse `err` later, in `historyEntry`. I don't really see the point of doing that in separate PR or commit.

aanm commented on cilium/cmd/map_get_history.go:
```diff
@@ -0,0 +1,79 @@
+}
+
+func printMapHistory(history *models.BPFMapHistory) {
+	w := tabwriter.NewWriter(os.Stdout, 5, 0, 3, ' ', 0)
```
And here, you would do a sort by time since the proposal that I've said with the `i++` would make the history to be out of order.

vadorovsky replied:
I think I will do sorting on the API side.

aanm reviewed (changes requested):
I would disable this option by default and only enable it once debug is also set.

vadorovsky committed:
option: Add options for BPF map operations history

This change introduces two option flags:

- map-history: which enables BPF map operations history
- map-history-max-entries: which sets the limit of entries

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
defaults: Add default BPF map operations history entries limit

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
daemon: Use BPF map operations history options

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>
